src/sema/eval.rs

```diff
@@ -196,3 +201,160 @@ pub fn eval_const_rational(
+                    Box::new(left),
+                    Box::new(right),
+                )
+            }
+        }
+
+        Expression::ShiftLeft(loc, ty, left, right) => {
+            let left = eval_constants_in_expression(left, ns);
+            let right = eval_constants_in_expression(right, ns);
+
+            if let (Expression::NumberLiteral(_, _, left), Expression::NumberLiteral(_, _, right)) =
+                (&left, &right)
+            {
+                overflow_check(ns, left.shl(right.to_u32().unwrap()), ty.clone(), *loc);
+                bigint_to_expression(loc, ty, left.shl(right.to_u32().unwrap()))
+            } else {
+                Expression::ShiftLeft(*loc, ty.clone(), Box::new(left), Box::new(right))
+            }
+        }
+        _ => expr.clone(),
+    }
+}
+
+fn bigint_to_expression(loc: &Loc, ty: &Type, n: BigInt) -> Expression {
+    let n = match ty {
+        Type::Uint(bits) => {
+            if n.bits() > *bits as u64 {
+                let (_, mut bs) = n.to_bytes_le();
+                bs.truncate(*bits as usize / 8);
+
+                BigInt::from_bytes_le(Sign::Plus, &bs)
+            } else {
+                n
+            }
+        }
+        Type::Int(bits) => {
+            if n.bits() > *bits as u64 {
+                let mut bs = n.to_signed_bytes_le();
+                bs.truncate(*bits as usize / 8);
+
+                BigInt::from_signed_bytes_le(&bs)
+            } else {
+                n
+            }
+        }
+        Type::StorageRef(..) => n,
+        _ => unreachable!(),
+    };
+
+    Expression::NumberLiteral(*loc, ty.clone(), n)
+}
+
+fn overflow_check(ns: &mut Namespace, result: BigInt, ty: Type, loc: Loc) {
+    if let Type::Uint(bits) = ty {
+        // If the result sign is minus, throw an error.
+        if let Sign::Minus = result.sign() {
+            ns.diagnostics.push(Diagnostic::error(
+                loc,
+            format!( "Type int_const {:?} is not implicitly convertible to expected type {:?}. Cannot implicitly convert signed literal to unsigned type.",result,ty),
+            ));
+        }
+
+        // If bits of the result is more than bits of the type, throw and error.
+        if result.bits() > bits as u64 {
+            ns.diagnostics.push(Diagnostic::error(
+                loc,
+                format!("Type int_const {:?} is not implicitly convertible to expected type {:?}. Literal is too large to fit in {:?}.",result,ty,ty),
+            ));
+        }
+    }
+
+    if let Type::Int(bits) = ty {
+        // If number of bits is more than what the type can hold. BigInt.bits() is not used here since it disregards the sign.
+        if result.to_signed_bytes_be().len() * 8 > (bits as usize) {
+            ns.diagnostics.push(Diagnostic::error(
+                loc,
+                format!("Type int_const {:?} is not implicitly convertible to expected type {:?}. Literal is too large to fit in {:?}.",result,ty,ty),
```

Comment:
`"value {} does not fit into type {}"`

Is tested with `int8 x = -128;` or `int16 y = -32768;`